File: autotoy/src/car/src/controller.cpp
#include "ros/ros.h"
#include "track/Line.h"
#include "car/Control.h"
#include "car/Location.h"
#include "car/Velocity.h"
#include <vector>

using namespace std;


class Controller {
  
  ros::NodeHandle nodeHandler;
  ros::Publisher publisher;
  ros::Subscriber targetLineSub;
  ros::Subscriber locationSub;
  ros::Subscriber velocitySub;
  
  public:
    float targetLineAngle, headingAngle, velocity, velocityRef, acceleration, yawRate;
    
    // Constructor
    Controller() {      
      targetLineAngle, headingAngle, velocity, yawRate, acceleration = 0;
      velocityRef = 3; // m/s
      
      // Inititate publisher and subscribers
      publisher = nodeHandler.advertise<car::Control>("car/controls", 1000);
      targetLineSub = nodeHandler.subscribe("car/targetline", 1000, &Controller::targetLineCallback, this);
      locationSub = nodeHandler.subscribe("car/location", 1000, &Controller::locationCallback, this); 
      velocitySub = nodeHandler.subscribe("car/velocity", 1000, &Controller::velocityCallback, this);
    };
    
    void talker() {
      float headingError = targetLineAngle - headingAngle;
      float velocityError = velocityRef - velocity;
      
      // Proportional control
      float Kp_yaw = 1.;
      float Kp_a = 0.5;
      yawRate = Kp_yaw * headingError;
      acceleration = Kp_a * velocityError;
            
      car::Control message;
      message.acceleration = acceleration;
      message.yawrate = yawRate;            
      publisher.publish(message);
    }
    
    void targetLineCallback(const track::Line& message) {
      ROS_INFO("Received target line data");
      
      if (message.points.size() > 0) {
        vector<float> currentPoint = {message.points[0].x, message.points[0].y};
        vector<float> nextPoint = {message.points[1].x, message.points[1].y};
        
        float angle = atan((nextPoint[1] - currentPoint[1]) / (nextPoint[0] - currentPoint[0] + 0.001));
        if (angle < 0) {
          angle += 180;
        }
        ROS_ERROR("%f", nextPoint[1]);
        
        targetLineAngle = angle;
      }
      
      talker();
    }
    
    void locationCallback(const car::Location& message) {
      ROS_INFO("Received location data");
      
      headingAngle = message.heading;
    }  
    
    void velocityCallback(const car::Velocity& message) {
      ROS_INFO("Received current velocity");
      
      velocity = message.velocity;
    }
  
};


int main(int argc, char **argv) { 
  
  // Initiate ros node
  ros::init(argc, argv, "controller");
  
  // Instantiate Controller class
  Controller controller;
  
  ros::spin();
  
  return 0;
}

File: autotoy/src/car/src/controller.py
#!/usr/bin/env python
import rospy
import math
from track.msg import Line
from car.msg import Control, Location

class Controller:
    
    def __init__(self):
        self.target_line_angle = 0
        self.heading_angle = 0
        
        # Initialise node
        rospy.init_node('controller')
        
        # Inititate publisher and subscribers
        self.publisher = rospy.Publisher('/car/controls', Control, queue_size=10)
        try:
            self.listener()
        except rospy.ROSInterruptException:
            pass
        
    def listener(self):        
        rospy.Subscriber('/car/targetline', Line, self.target_line_callback)
        rospy.Subscriber('/car/location', Location, self.location_callback)
        
        print('Listening')
        rospy.spin()
        
    def talker(self):    
        error = self.target_line_angle - self.heading_angle
        
        # Convert error to yaw rate
        # PID stuff...
        acceleration = 4
        yaw_rate = 1
        
        self.publisher.publish(acceleration, yaw_rate)        
              
    def target_line_callback(self, data):
        print('Received target line data')
        
        current_point = (data.points[0].x, data.points[0].y)
        next_point = (data.points[1].x, data.points[1].y)
        
        angle = math.atan((next_point[1] - current_point[1]) / (next_point[0] - current_point[0]))
        if angle < 0: 
            angle += 180
             
        self.target_line_angle = angle
    
        self.talker()
    
    def location_callback(self, data):
        print('Received location and heading')
        
        # Handle location and heading data
        self.heading_angle = data.heading


if __name__ == '__main__':    
    Controller()
    